### PR TITLE
[PyTorch][Edge] Serializing sub modules with same names

### PR DESCRIPTION
**Pull request by pavithranrao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61933

### Issue:

SubModules with same name are not serialized correctly in bytecode format while using `_save_for_mobile`. These submodules are not distinguished as different modules even though they have different foward, setstate etc if they have the same name.

### Fix:
Mangler creates unique names so that modules and submodules that have same names can be uniquely identified  while saving the module. @myuan rightly pointed out the underlying issue that mangler is not used in the process of saving bytecode and hence unique references for the submodules are not created. Please refer to the notebook to repro the issue: N777224

### Diff:
The above idea of fix is implemented. The mangled names are used in bytecode thereby the files in `code/` directory now have right reference to the `bytecode.pkl`

Will this have backward compatibility?
@myuan please feel free to correct or update this.
Yes. This fix impacts only modules with same name sub modules which were not serialized correctly before. Existing modules should have correct references and `_load_for_mobile` must not see any change. To confirm this the existing test cases need to pass for the diff to be approved and shipped.

Differential Revision: [D29669831](https://our.internmc.facebook.com/intern/diff/D29669831/)

### CODE REVIEW

**Review by iseeyuan — APPROVED** [review]
LGTM. Next time will do the reviews here instead of in internal diff.

**Comment by kimishpatel** [comment]
Thanks for the detailed summary. Some CI jobs are failing. Make sure to take a look

**Comment by pavithranrao** [comment]
@iseeyuan The error in CI build is:

```
test_backend.cpp:(.text+0x4f0): undefined reference to `torch::jit::mobile::Function::qualname() const'
```

But not sure why this happens? We have `const c10::QualifiedName& qualname() const` in `/torch/csrc/jit/mobile/function.h`

https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/function.h#L21

**Comment by kimishpatel** [comment]
> @iseeyuan, @kimishpatel The error in CI build is:
> 
> ```
> test_backend.cpp:(.text+0x4f0): undefined reference to `torch::jit::mobile::Function::qualname() const'
> ```
> 
> But not sure why this happens? We have `const c10::QualifiedName& qualname() const` in `/torch/csrc/jit/mobile/function.h`
> 
> https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/function.h#L21

This may need to be marked with TORCH_API, but I dont think we should have `torch::jit::mobile::Function` be exposed. Is there any other way to write the test. Let me take a look.

**Comment by kimishpatel** [comment]
> @iseeyuan, @kimishpatel The error in CI build is:
> 
> ```
> test_backend.cpp:(.text+0x4f0): undefined reference to `torch::jit::mobile::Function::qualname() const'
> ```
> 
> But not sure why this happens? We have `const c10::QualifiedName& qualname() const` in `/torch/csrc/jit/mobile/function.h`
> 
> https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/function.h#L21

BTW for context, this is a link time error. So at linking time it is trying to find where the qualname function is defined. We likely link test_backend against libtorch and since this function/class is not marked with TORCH_API its visiblity is hidden such that it is marked for internal linkage only, that is it can be used by some other code that gets compiled as part of libtorch only. https://gcc.gnu.org/wiki/Visibility. I havent read though.

**Comment by pavithranrao** [comment]
@kimishpatel thanks for explaining that. Got confused as it ran in internal successfully, where in OSS CI it fails. I didnt know tests are run on the built `libtorch` instead of the code. So what do we do in this case? Do we disable this test in OSS or expose a function API in `Function.h` or do we have a better way?

File: torch/csrc/jit/serialization/export_module.cpp [hunk 1]
```diff
@@ -60,7 +60,9 @@ static IValue Table(
 std::pair<IValue, IValue> getFunctionTuple(
     const Module& module,
     const Function& func,
-    BackendDebugInfoRecorder& debug_info_recorder) {
+    BackendDebugInfoRecorder& debug_info_recorder,
+    const std::basic_string<char>& qn,
+    TypeNameUniquer& type_name_uniquer_) {
```
Comment (kimishpatel) [inline torch/csrc/jit/serialization/export_module.cpp:65]: BTW trailing _ is for private class member names.

Comment (pavithranrao) [reply torch/csrc/jit/serialization/export_module.cpp:65]: We are using a const reference of the `type_name_uniquer_` from `ScriptModuleSerializer` in these functions. I thought will have the same name making it easier to search and understand. Anyone who searches for `type_name_uniquer_` in this file can discover the usages. I understand this violates the naming convention of the variable. What do you recommend in this case?

Comment (kimishpatel) [reply torch/csrc/jit/serialization/export_module.cpp:65]: I think that’s a good reason but you can leave name with _ in a comment so if someone greps they can find it which is what I presume you want.

**Comment by kimishpatel** [comment]
Btw regarding test in oss I suppose you can also just read bytecode via show_pickle util or just grepping for string in bytecode.pkl?

**Pull request closed without merging**